File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081334Z_claude_claude-sonnet-5_php-symfony/src/Audit/AuditController.php
<?php

namespace App\Audit;

use App\Storage\Database;
use App\Support\Json;
use Symfony\Component\HttpFoundation\JsonResponse;

/** Handlers for campaign audit-log and export endpoints (/v1/campaigns/{id}/audit, /export). */
final class AuditController
{
    public function audit(string $campaignId): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT id FROM campaigns WHERE id = ?');
        $stmt->execute([$campaignId]);
        if ($stmt->fetchColumn() === false) {
            return Json::error('campaign not found', 404);
        }

        $count = static function (string $table) use ($db, $campaignId): int {
            $stmt = $db->prepare("SELECT COUNT(*) FROM {$table} WHERE campaign_id = ?");
            $stmt->execute([$campaignId]);
            return (int) $stmt->fetchColumn();
        };

        return new JsonResponse([
            'campaign_id' => $campaignId,
            'events' => $count('campaign_events'),
            'quests' => $count('campaign_quests'),
            'npcs' => $count('campaign_npcs'),
            'sessions' => $count('campaign_sessions'),
        ]);
    }

    public function export(string $campaignId): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT id, name FROM campaigns WHERE id = ?');
        $stmt->execute([$campaignId]);
        $campaign = $stmt->fetch(\PDO::FETCH_ASSOC);
        if ($campaign === false) {
            return Json::error('campaign not found', 404);
        }

        $count = static function (string $table) use ($db, $campaignId): int {
            $stmt = $db->prepare("SELECT COUNT(*) FROM {$table} WHERE campaign_id = ?");
            $stmt->execute([$campaignId]);
            return (int) $stmt->fetchColumn();
        };

        return new JsonResponse([
            'campaign_id' => $campaign['id'],
            'name' => $campaign['name'],
            'characters' => $count('campaign_characters'),
            'quests' => $count('campaign_quests'),
            'npcs' => $count('campaign_npcs'),
            'inventory_items' => $count('campaign_inventory'),
            'sessions' => $count('campaign_sessions'),
            'schema_version' => Database::SCHEMA_VERSION,
        ]);
    }
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051117Z_claude_claude-sonnet-5_ruby-rails/app/controllers/audit_controller.rb
# Campaign audit log and deterministic export summaries. Reads only;
# aggregates counts from the campaign hash maintained by the other
# domain controllers.
class AuditController < ApplicationController
  def audit_log
    campaign = CAMPAIGNS[params[:campaign_id]]
    if campaign.nil?
      render json: { error: 'campaign not found' }, status: :not_found
      return
    end

    render json: {
      campaign_id: params[:campaign_id],
      events: (campaign[:events] || []).length,
      quests: (campaign[:quests] || {}).length,
      npcs: (campaign[:npcs] || {}).length,
      sessions: (campaign[:sessions] || {}).length
    }
  end

  def export_campaign
    campaign = CAMPAIGNS[params[:campaign_id]]
    if campaign.nil?
      render json: { error: 'campaign not found' }, status: :not_found
      return
    end

    render json: {
      campaign_id: params[:campaign_id],
      name: campaign[:name],
      characters: (campaign[:characters] || []).length,
      quests: (campaign[:quests] || {}).length,
      npcs: (campaign[:npcs] || {}).length,
      inventory_items: (campaign[:inventory] || []).length,
      sessions: (campaign[:sessions] || {}).length,
      schema_version: 1
    }
  end
end
